File: src/api/timetable.ts
import { TabLocation, type SessionHandle, type Timetable } from "~/models";
import { propertyCaseInsensitive } from "./private/property-case-insensitive";
import { RequestFN } from "~/core/request-function";
import { encodeUserResource } from "~/encoders/user-resource";
import { decodeTimetable } from "~/decoders/timetable";
import { encodePronoteDate } from "~/encoders/pronote-date";

const timetable = async (session: SessionHandle, index: number, additional = {}): Promise<Timetable> => {
  const request = new RequestFN(session, "PageEmploiDuTemps", {
    _Signature_: { onglet: TabLocation.Timetable },

    donnees: {
      estEDTAnnuel: false,
      estEDTPermanence: false,

      avecAbsencesEleve: false,
      avecRessourcesLibrePiedHoraire: false,

      avecAbsencesRessource: true,
      avecInfosPrefsGrille: true,
      avecConseilDeClasse: true,
      avecCoursSortiePeda: true,
      avecDisponibilites: true,
      avecRetenuesEleve: true,

      edt: { G: 16, L: "Emploi du temps" },

      ...propertyCaseInsensitive("ressource", encodeUserResource(session.user.resources[index])),
      ...additional
    }
  });

  const response = await request.send();
  return decodeTimetable(response.data.donnees, session);
};

export const timetableFromWeek = async (session: SessionHandle, weekNumber: number, index = 0): Promise<Timetable> => {
  return timetable(session, index, propertyCaseInsensitive("numeroSemaine", weekNumber));
};

export const timetableFromIntervals = async (session: SessionHandle, startDate: Date, endDate?: Date, index = 0): Promise<Timetable> => {
  return timetable(session, index, {
    ...propertyCaseInsensitive("dateDebut", {
      _T: 7,
      V: encodePronoteDate(startDate)
    }),

    ...(endDate && propertyCaseInsensitive("dateFin", {
      _T: 7,
      V: encodePronoteDate(endDate)
    }))
  });
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/timetable.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.api.private.propertyCaseInsensitive
import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeTimetable
import ink.literate.pawnote.encoders.encodePronoteDate
import ink.literate.pawnote.encoders.encodeUserResource
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import ink.literate.pawnote.models.Timetable
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun timetable(
    session: SessionHandle,
    index: Int,
    additional: JsonObject = buildJsonObject {}
): Timetable {
  val request =
      RequestFN(
          session.information,
          "PageEmploiDuTemps",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Timetable.code) }

                put(
                    "donnees",
                    JsonObject(
                        buildJsonObject {
                          put("estEDTAnnuel", false)
                          put("estEDTPermanence", false)

                          put("avecAbsencesEleve", false)
                          put("avecRessourcesLibrePiedHoraire", false)

                          put("avecAbsencesRessource", true)
                          put("avecInfosPrefsGrille", true)
                          put("avecConseilDeClasse", true)
                          put("avecCoursSortiePeda", true)
                          put("avecDisponibilites", true)
                          put("avecRetenuesEleve", true)

                          putJsonObject("edit") {
                            put("G", 16)
                            put("L", "Emploi du temps")
                          }
                        } +
                            propertyCaseInsensitive(
                                "ressource", encodeUserResource(session.user.resources[index])) +
                            additional))
              }))

  val response = request.send()
  return decodeTimetable(
      Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject, session.instance)
}

suspend fun timetableFromWeek(session: SessionHandle, weekNumber: Int, index: Int = 0): Timetable {
  return timetable(
      session,
      index,
      propertyCaseInsensitive("numeroSemaine", Json.parseToJsonElement(weekNumber.toString())))
}

suspend fun timetableFromIntervals(
    session: SessionHandle,
    startDate: LocalDateTime,
    endDate: LocalDateTime? = null,
    index: Int = 0
): Timetable {
  return timetable(
      session,
      index,
      JsonObject(
          propertyCaseInsensitive(
              "dateDebut",
              buildJsonObject {
                put("_T", 7)
                put("V", encodePronoteDate(startDate))
              }) +
              if (endDate != null)
                  propertyCaseInsensitive(
                      "dateFin",
                      buildJsonObject {
                        put("_T", 7)
                        put("V", encodePronoteDate(endDate))
                      })
              else buildJsonObject {}))
}
